community/index/src/main/java/org/neo4j/index/gbptree/SeekCursor.java

```diff
@@ -44,7 +43,102 @@
+ * <p>
+ * <strong>Case 1 - Split in current node</strong>
+ * <p>
+ * <em>Forward seek</em>
+ * <p>
+ * Here, the seeker is seeking forward and has read K0, K1, K2 and is about to read K3.
+ * Then K4 is suddenly inserted and a split happens.
+ * Seeker will now wake up and find that he is now outside the range of (previously returned key, end of range).
+ * But he knows that he can continue to read forward until he hits the previously returned key, which is K2 and
+ * then continue to return the next key, K3.
+ * <pre>
+ *     Seeker->
+ *        v
+ * [K0 K1 K2 K3]
+ *
+ *     Seeker->
+ *        v
+ * [K0 K1 __ __]<->[K2 K3 K4 __]
+ * </pre>
+ * <em>Backward seek</em>
+ * <p>
+ * Here, the seeking is seeking backwards and has only read K3 and is about to read K2.
+ * Again, K4 is inserted, causing the same split as above.
+ * Seeker now wakes up and find that the next key he can return is K1. What he do not see is that K2 has been
+ * moved to the previous sibling and so, because he can not find the previously returned key, K3, in the current
+ * node and because the next to return, K1, is located to the far right in the current node he need to jump back to
+ * the previous sibling to find where to start again.
+ * <pre>
+ *      <-Seeker
+ *           v
+ * [K0 K1 K2 K3]
+ *
+ *      <-Seeker (WRONG)
+ *           v
+ * [K0 K1 __ __]<->[K2 K3 K4 __]
+ *
+ *                <-Seeker (RIGHT)
+ *                     v
+ * [K0 K1 __ __]<->[K2 K3 K4 __]
+ * </pre>
+ * <p>
+ * <strong>Case 2 - Split in next node while seeker is moving to that node</strong>
+ * <p>
+ * <em>Forward seek</em>
+ * <p>
+ * Seeker has read K0, K1 on node 1  and has just moved to right sibling, node 2.
+ * Now, K6 is inserted and a split  happens in node 2, right sibling of node 1.
+ * Seeker wakes up and continues to read on node 2. Everything is fine.
+ * <pre>
+ *                   Seeker->
+ *                      v
+ * 1:[K0 K1 __ __]<->2:[K2 K3 K4 K5]
+ *
+ *                   Seeker->
+ *                      v
+ * 1:[K0 K1 __ __]<->2:[K2 K3 __ __]<->3:[K4 K5 K6 __]
+ * </pre>
+ * <em>Backward seek</em>
+ * <p>
+ * Seeker has read K4, K5 and has just moved to left sibling, node 1.
+ * Insert K6, split in node 2. Note that keys are move to node 3, passed our seeker.
+ * Seeker wakes up and see K1 as next key but misses K3 and K2.
+ * <pre>
+ *       <--Seeker
+ *             v
+ * 1:[K0 K1 K2 K3]<->2:[K4 K5 __ __]
+ *
+ *        <-Seeker
+ *             v
+ * 1:[K0 K1 __ __]<->3:[K2 K3 K4 __]<->2:[K5 K6 __ __]
+ * </pre>
+ * To guard for this, seeker 'scout' next sibling before moving there and read first key that he expect to see, K3
+ * in this case. By using a linked cursor to 'scout' we create a consistent read over the node gap. If there us
+ * suddenly another key when he goes there he knows that he could have missed some keys and he needs to go back until
+ * he find the place where he left off, K4.
```

Comment:
Now this cursor is almost impossible to read and understand even when i can grab @burqen and bomb him with unstoppable rain of questions. Have you guys consider possibility to split this cursor at least into seek cursor and then have 2 actual public visible forward and backward cursor classes that do the stuff only for their use case. That at least will gonna simplify logic around everything that responsible for the navigation logic